lib/types.nix

```diff
@@ -113,8 +113,20 @@ rec {
       name
     , # Description of the type, defined recursively by embedding the wrapped type if any.
       description ? null
-    , # Function applied to each definition that should return true if
+    , # ☠ DO NOT USE WITHOUT KNOWING WHAT YOU ARE DOING ☠
+      # Function applied to each definition that should return true if
       # its type-correct, false otherwise.
+      # However, for composite types (lists, attrsets, submodules), this only checks
+      # the top level (i.e. is this a list) without recursing into the elements.
+      # This is done because the inner elements are checked differently by the module
+      # system, at the cost of making this function useless as a pure type check.
```

Comment:
```suggestion
      # Function applied to each definition that must return false when a definition
      # does not match the type. It should not check more than the root of the value,
      # because evaluating contained reduces laziness, leading to unnecessary
      # infinite recursions in the module system.
      # Further checks of contained values should be performed by throwing in
      # the merge function.
      # Strict and deep type checking can be performed by calling lib.deepSeq on
      # the merged value.
```